src/renderer/renderer.cpp

```diff
@@ -0,0 +1,137 @@
+#include "renderer.hpp"
+
+#include "common.hpp"
+#include "file_manager.hpp"
+#include "fonts/fonts.hpp"
+#include "gui.hpp"
+#include "pointers.hpp"
+
+#include <backends/imgui_impl_dx11.h>
+#include <backends/imgui_impl_win32.h>
+#include <imgui.h>
+#include <imgui_internal.h>
+
+IMGUI_IMPL_API LRESULT ImGui_ImplWin32_WndProcHandler(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam);
+
+namespace big
+{
+	renderer::renderer() :
+	    m_font_mgr()
+	{
+		
+	}
+
+	bool renderer::init()
+	{
+		if (!g_pointers->m_gta.m_swapchain || !*g_pointers->m_gta.m_swapchain)
+		{
+			LOG(FATAL) << "Invalid swapchain ptr";
+
+			return false;
+		}
+		m_dxgi_swapchain = *g_pointers->m_gta.m_swapchain;
+
+		if (m_dxgi_swapchain->GetDevice(__uuidof(ID3D11Device), reinterpret_cast<void**>(&m_d3d_device)) < 0)
+		{
+			LOG(FATAL) << "Failed to get D3D device.";
+
+			return false;
+		}
+		m_d3d_device->GetImmediateContext(&m_d3d_device_context);
+
+		auto file_path = g_file_manager.get_project_file("./imgui.ini").get_path();
+
+		ImGuiContext* ctx = ImGui::CreateContext();
+
+		static std::string path = file_path.make_preferred().string();
+		ctx->IO.IniFilename     = path.c_str();
+
+		ImGui_ImplDX11_Init(m_d3d_device, m_d3d_device_context);
+		ImGui_ImplWin32_Init(g_pointers->m_hwnd);
+
+		return true;
+	}
+
+	void renderer::destroy()
+	{
+		ImGui_ImplWin32_Shutdown();
+		ImGui_ImplDX11_Shutdown();
+		ImGui::DestroyContext();
+	}
+
+	bool renderer::add_dx_callback(dx_callback callback, uint32_t priority)
+	{
+		if (!m_dx_callbacks.insert({priority, callback}).second)
+		{
+			LOG(WARNING) << "Duplicate priority given on DX Callback!";
+
+			return false;
+		}
+		return true;
+	}
+
+	void renderer::add_wndproc_callback(wndproc_callback callback)
+	{
+		m_wndproc_callbacks.emplace_back(callback);
+	}
+
+	void renderer::on_present()
+	{
+		if (m_font_mgr.rebuild_required())
+		{
+			m_font_mgr.rebuild_now();
+			pre_reset();
+			post_reset();
```

Comment:
These are pretty heavy, @xiaoxiao921 any ideas on doing it with more lightweight calls?